src/text/parsers/boolean.rs

```diff
@@ -1,16 +1,17 @@
 //! Parsing logic for the text representation of boolean values.
 
+use crate::text::parse_result::IonParseResult;
 use nom::bytes::streaming::tag;
 use nom::combinator::{map, recognize};
 use nom::sequence::terminated;
-use nom::{IResult, Parser};
+use nom::Parser;
 
 use crate::text::parsers::stop_character;
 use crate::text::text_value::TextValue;
 
 /// Matches the text representation of a boolean value and returns the resulting `true` or `false`
 /// as a [TextValue::Boolean].
-pub(crate) fn parse_boolean(input: &str) -> IResult<&str, TextValue> {
+pub(crate) fn parse_boolean(input: &str) -> IonParseResult<TextValue> {
     map(
```

Comment:
Great question! All of `nom`'s parsers implement [the `Parser` trait](https://docs.rs/nom/latest/nom/trait.Parser.html), which looks like this:

```rust
pub trait Parser<I, O, E> {
  // Input -----^  ^  ^
  // Output -------+  |
  // Error -----------+

  // Takes an input (`I`) and returns a `Result` with either:
  // * an `Ok((I, O))` tuple with the remaining (unmatched) input and the `O` output value it parsed
  // * an `Err(I, nom::Err<E>)` with the input it couldn't match and the error `E` wrapped in an `Err`
  fn parse(&mut self, input: I) -> IResult<I, O, E>;
  
  /// ...other stuff
}
```

For any type that implements `Parser`, they'll need to pick their `I`, `O`, and `E` types. 

For us, `I` is always `&str`. `O` depends on what the parser's trying to do, and `E` is either `nom::error::Error` (the default error type for every core `nom` parser) or `IonParseError` (for parser functions we wrote ourselves).

The compiler tries to figure out which `E` to use based on how the parser is called. For [`map`](https://docs.rs/nom/latest/nom/combinator/fn.map.html), this is easy. Here's its function signature:
```rust
pub fn map<I, O1, O2, E, F, G>(mut parser: F, mut f: G) -> impl FnMut(I) -> IResult<I, O2, E>
where                  //                                                                  ^
  F: Parser<I, O1, E>, // <--- This is the only restriction on `E`,                        |
                       //      and the same `E` is used in the function's return type -----+
  G: FnMut(O1) -> O2,
```

The compiler will say that `map`'s `E` type is whatever `parser`'s `E` type is. Other parsers are tougher because they take _multiple_ parsers as parameters. For example, consider [the `pair` parser](https://docs.rs/nom/latest/nom/sequence/fn.pair.html):

```rust
pub fn pair<I, O1, O2, E: ParseError<I>, F, G>(
  mut first: F,
  mut second: G,
) -> impl FnMut(I) -> IResult<I, (O1, O2), E>
where
  F: Parser<I, O1, E>, // These parsers can produce different output types (`O1` and `O2`),
  G: Parser<I, O2, E>, // but they must return the *same* error type.
```

In situations like this, we can use `upgrade()` on the core `nom` parsers to change their `E` type to be `IonParseError`. In that way, all of the parsers being passed as parameters will have the same error type.

Other links:
* [the `IResult` type alias](https://docs.rs/nom/latest/nom/type.IResult.html)
* [the `nom::Err<E>` enum](https://docs.rs/nom/latest/nom/enum.Err.html)